src/SubmitCarHandler.php

```diff
@@ -284,6 +290,33 @@ public function view_preview() {
 	}
 
 	/**
+	 * Process a new listing, called after the view_preview() method
+	 */
+	public function process_new_listing() {
+
+		// vehicle ID that is requested to publish
+		$vehicle_id = $this->listing_id;
+
+		// redirect user to login screen if they can't edit this listing
+		if ( ! wp_car_manager()->service( 'user_manager' )->can_edit_listing( $vehicle_id ) ) {
+			wp_redirect( wp_login_url( add_query_arg( 'wpcm_publish', $vehicle_id ) ) );
+			exit;
+		}
+
+		// get vehicle object
+		/** @var Vehicle $vehicle */
+		$vehicle = wp_car_manager()->service( 'vehicle_factory' )->make( $vehicle_id );
```

Comment:
The following class that extends Pimple may be useful to you.

- https://github.com/ibericode/boxzilla-wp/blob/master/src/di/class-container-with-property-access.php

It allows you to typehint properties on your service container, like this.

- https://github.com/ibericode/boxzilla-wp/blob/master/src/class-boxzilla.php